Add support for Xiaomi Mi Note 10 (tucana)

SanyaPilot opened the pull request:
- Adapted STMicroelectronics FTS touchscreen driver
- Adapted AW3644 LEDs driver
- Renamed Xiaomi F4-41-04-0A driver to Visionox G2647FB105
- Fixed device tree

SanyaPilot committed:
arm64: dts: qcom: sm7150: tucana: Fixups and improvements

- Fix reserved memory regions
- Add panel node (probably Visionox one, Tianma is unsupported for now) and fixed regulators for it
- Add FTS touchscreen support, add a fixed regulator for it
- Add battery design capacity
- Add sound support (only main speaker is supported for now)
- Add QCOM flash support (two LEDs)
- Add soft light LEDs support (aw3644, two LEDs)
- Add haptics support (drv2605l)

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

SanyaPilot committed:
drivers: input: Add STMicroelectronics FTS touchscreen driver

This driver was published at https://github.com/STMicroelectronics/STTouchController-Android-Drivers

Modifications:
- Adapt for changes in recent kernel versions
- Allow building as a module (fts521)
- Add MODULE_DEVICE_TABLE macros to support autoloading
- Parse configuration data from a DT node
- Use a threaded IRQ handler instead of a blocking one (as in Xiaomi's changes)
- Set the required voltage before enabling AVDD regulator
- Add pinctrl support (as in Xiaomi's changes), rename them to "default" and "sleep"
- Fix various warnings (not all)
- Turn down log verbosity a much
- clang-formatted the entire driver
- Remove /proc file for debugging
- Remove the firmware updater (there are not going to be future updates)
- Remove production tests, move IC initialization function code to fts.c. Why this even exists in a driver...
- Remove fts_limits.h (some data for production tests) and fts_fw.h (firmware)
- Remove various code that became unused

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

SanyaPilot committed:
drivers: leds: Import AW3644 driver from Xiaomi's changes

Modifications:
- Move leds-aw3644.h contents here
- Texas Instruments -> Awinic (???)
- Adapt for the latest kernel
- Set MODULE_DEVICE_TABLE macro
- Parse torch LEDs from device tree, drop hardcoded LED names, they are assigned automatically by the kernel now
- Fix warnings
- clang-formatted the entire driver

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

SanyaPilot committed:
arm64: configs: sm7150: Enable drivers for tucana and toco

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

JIaxyga commented on drivers/gpu/drm/panel/Kconfig:
```diff
@@ -961,6 +961,16 @@ config DRM_PANEL_TRULY_NT35597_WQXGA
+	depends on OF
+	depends on DRM_MIPI_DSI
+	depends on BACKLIGHT_CLASS_DEVICE
+	select VIDEOMODE_HELPERS
```
drop this line

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -287,9 +287,9 @@ static int xiaomi_f4_41_06_0a_probe(struct mipi_dsi_device *dsi)
+	struct visionox_g2647fb105 *ctx = mipi_dsi_get_drvdata(dsi);
 	int ret;
 
 	ret = mipi_dsi_detach(dsi);
```
drop
```
	int ret;

	ret = mipi_dsi_detach(dsi);
	if (ret < 0)
		dev_err(&dsi->dev, "Failed to detach from DSI host: %d\n", ret);
```

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -205,7 +205,7 @@ static int xiaomi_f4_41_06_0a_bl_update_status(struct backlight_device *bl)
 
 // TODO: Check if /sys/class/backlight/.../actual_brightness actually returns
```
Check please

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -287,9 +287,9 @@ static int xiaomi_f4_41_06_0a_probe(struct mipi_dsi_device *dsi)
 	return 0;
```
switch to devm_mipi_attach:
```
	ret = devm_mipi_dsi_attach(dev, dsi);
	if (ret < 0) {
		drm_panel_remove(&ctx->panel);
		return dev_err_probe(dev, ret, "Failed to attach to DSI host\n");
	}
```

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -238,13 +238,13 @@ xiaomi_f4_41_06_0a_create_backlight(struct mipi_dsi_device *dsi)
 	};
```
.brightness = 1023,
.max_brightness = 2047,

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -13,28 +13,28 @@
 	struct drm_panel panel;
 	struct mipi_dsi_device *dsi;
 	struct gpio_desc *reset_gpio;
 	bool prepared;
```
Drop this line

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -106,18 +106,18 @@ static int xiaomi_f4_41_06_0a_off(struct xiaomi_f4_41_06_0a *ctx)
 	struct device *dev = &ctx->dsi->dev;
 	int ret;
 
 	if (ctx->prepared)
```
And this too

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -128,16 +128,16 @@ static int xiaomi_f4_41_06_0a_prepare(struct drm_panel *panel)
 	struct device *dev = &ctx->dsi->dev;
 	int ret;
 
 	if (!ctx->prepared)
```
And everything related to "prepared". I might have missed something

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts:
```diff
@@ -28,40 +37,44 @@
-			reg = <0x0 0x95100000 0x0 0x2600000>;
-			no-map;
-		};
+	vreg_touch_vdd: regulator-touch-vdd {
```
Here the sorting is alphabetical, so move after reserved-memory

JIaxyga commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -299,22 +299,22 @@ static void xiaomi_f4_41_06_0a_remove(struct mipi_dsi_device *dsi)
-static const struct of_device_id xiaomi_f4_41_06_0a_of_match[] = {
-	{ .compatible = "xiaomi,f4-41-06-0a" }, // FIXME
+static const struct of_device_id visionox_g2647fb105_of_match[] = {
+	{ .compatible = "visionox,g2647fb105" },
```
you need to fix dts toco immediately, but in a separate commit

SanyaPilot committed:
drivers: gpu: drm: panel: Add Visionox G2647FB105 panel driver

Add the driver for Visionox G2647FB105 6.47" FHD Plus CMD mode OLED panel
support found in Xiaomi Mi Note 10 / Mi CC9 Pro (sm7150-xiaomi-tucana) and
Xiaomi Mi Note 10 Lite (sm7150-xiaomi-toco)

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

SanyaPilot committed:
dt-bindings: display: panel: Add Visionox G2647FB105

The Visionox G2647FB105 is a 6.47 inch 1080x2340 MIPI-DSI CMD mode OLED panel
found in Xiaomi Mi Note 10 / Mi CC9 Pro (sm7150-xiaomi-tucana) and
Xiaomi Mi Note 10 Lite (sm7150-xiaomi-toco)

Add a dt-binding for it.

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

SanyaPilot committed:
arm64: dts: qcom: sm7150: toco: Fixup panel node

New panel driver has appeared, so fix device tree

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

Gelbpunkt commented on arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts:
```diff
@@ -85,23 +116,81 @@
+		fts,irq-gpio = <&tlmm 9 GPIO_TRANSITORY>;
+		fts,reset-gpio-enable;
+		fts,reset-gpio = <&tlmm 8 GPIO_ACTIVE_HIGH>;
+		fts,irq-flags = <0x2008>; /* oneshot and trigger low */
```
`IRQF_ONESHOT | IRQF_TRIGGER_LOW`, see include/linux/interrupt.h

SanyaPilot replied:
I just can't include this header into the device tree, it gives me
```
../arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts:11:10: fatal error: linux/interrupt.h: No such file or directory
   11 | #include <linux/interrupt.h>
      |
```
Even if I manage to include it using an ugly `#include "../../../../../include/linux/interrupt.h"`, it fails bcoz of
```
../arch/arm64/boot/dts/qcom/../../../../../include/linux/interrupt.h:6:10: fatal error: linux/kernel.h: No such file or directory
    6 | #include <linux/kernel.h>
      |
```

SanyaPilot replied:
If I don't include it at all, compilation fails with
```
Error: ../arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts:171.20-21 syntax error
FATAL ERROR: Unable to parse input tree
```
Because it can't find those constants

Gelbpunkt commented on arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts:
```diff
@@ -85,23 +116,81 @@
 	status = "okay";
 
-	/* st,fts (touchscreen) @ 49 */
+	touchscreen: fts@49 {
+		compatible = "st,fts";
+		reg = <0x49>;
+		interrupt-parent = <&tlmm>;
+		interrupts = <9 IRQ_TYPE_LEVEL_LOW>;
+		pinctrl-0 = <&ts_int_active &ts_reset_active>;
+		pinctrl-1 = <&ts_int_suspend &ts_reset_suspend>;
+		pinctrl-names = "default", "sleep";
+		vdd-supply = <&vreg_touch_vdd>;
+		avdd-supply = <&vreg_l6c_3p0>;
+		fts,regulator_dvdd = "vdd";
+		fts,regulator_avdd = "avdd";
+		fts,irq-gpio = <&tlmm 9 GPIO_TRANSITORY>;
+		fts,reset-gpio-enable;
+		fts,reset-gpio = <&tlmm 8 GPIO_ACTIVE_HIGH>;
+		fts,irq-flags = <0x2008>; /* oneshot and trigger low */
+		fts,x-max = <1080>;
+		fts,y-max = <2340>;
+	};
```
This entire section could use some newlines

Gelbpunkt commented on drivers/gpu/drm/panel/panel-visionox-g2647fb105.c:
```diff
@@ -1,40 +1,43 @@
 // SPDX-License-Identifier: GPL-2.0-only
```
IMO it would be more appropriate to revert the old commit so that we can drop it nicely later.

Gelbpunkt reviewed (changes requested):
Looks good to me asides from one nit.

SanyaPilot committed:
Revert "drm: panel: Generate panel-xiaomi-f4-41-06-0a driver"

This reverts commit de81bf3ca3ca9a46076e005bed265b3e3207a937.

SanyaPilot committed:
arm64: configs: sm7150: Remove Xiaomi F4-41-04-0A panel driver

This driver was removed.

Signed-off-by: Alexander Baransky <alexander.baranskiy@yandex.ru>

JIaxyga commented:
thanks
